### PR TITLE
IBX-385: Fixed error handling when composer files are corrupted

### PR DESCRIPTION
**Pull request by konradoboza**
JIRA ticket: [IBX-385](https://issues.ibexa.co/browse/IBX-385)

>Currently, whenever composer.json or composer.lock files get accidentally corrupted, the system will return exceptions coming from ez-support-tools (where those files are parsed) which results in the system being unusable. The idea is to show some warnings in:
>- the dashboard,
> - the System Information tab,
> - ez-support-tools:dump-info
>if the composer files are either not accessible or don't contain valid data.

Depends on https://github.com/ezsystems/ezplatform-admin-ui/pull/1758.

TODO:
- [x] remove dependency to `ezplatform-admin-ui` PR before merge

//edit not sure to what I owe ezrobot's endorsement... 😄

### CODE REVIEW

File: Resources/translations/dashboard.en.xlf [hunk 1]
```diff
@@ -73,9 +73,9 @@
         <note>key: dashboard.ez_version.non_stable_ee</note>
       </trans-unit>
       <trans-unit id="c1fb5ba9fd547e30a49a4e1895822e69cd4a9dbc" resname="dashboard.ez_version.release_not_determined">
-        <source><![CDATA[The system could not find your <code>composer.lock</code> file. It's needed to determine information about
+        <source><![CDATA[The system could not find your <code>composer.lock</code> or <code>composer.json</code> file or it's not valid. It's needed to determine information about
```
Comment (gggeek) [inline Resources/translations/dashboard.en.xlf:76]: Nitpick: "or either is not valid. They are needed"

Comment (konradoboza) [reply Resources/translations/dashboard.en.xlf:76]: Let me ping an expert for her 5 cents 😄 @DominikaK could you please help with rephrasing this part?

File: SystemInfo/Collector/JsonComposerLockSystemInfoCollector.php [hunk 2]
```diff
@@ -74,6 +75,14 @@ public function collect()
         $lockData = json_decode(file_get_contents($this->lockFile), true);
         $jsonData = json_decode(file_get_contents($this->jsonFile), true);
 
+        if (!$lockData) {
```
Comment (adamwojs) [inline SystemInfo/Collector/JsonComposerLockSystemInfoCollector.php:78]:
```suggestion
        if ($lockData === null) {
```

> null is returned if the json cannot be decoded or if the encoded data is deeper than the nesting limit.

ref. https://www.php.net/manual/en/function.json-decode 

or

```suggestion
        if (!is_array($lockData)) {
```
as this is what `\EzSystems\EzSupportToolsBundle\SystemInfo\Collector\JsonComposerLockSystemInfoCollector::{extractPackages, extractRepositoryUrls}` expects as input

File: Resources/translations/dashboard.en.xlf (hunk 1 above)
Comment (DominikaK) [inline Resources/translations/dashboard.en.xlf:76]:
```suggestion
        <source><![CDATA[The system could not find your <code>composer.lock</code> or <code>composer.json</code> file, or one of them is invalid. The files are needed to determine information about
```
or, 
```suggestion
        <source><![CDATA[The system could not find your <code>composer.lock</code> or <code>composer.json</code> file, or the files are invalid. They are needed to determine information about
```

File: SystemInfo/Collector/JsonComposerLockSystemInfoCollector.php [hunk 3]
```diff
@@ -74,6 +75,14 @@ public function collect()
...
+            throw new Exception\ComposerFileValidationException($this->lockFile);
+        }
+
+        if (!$jsonData) {
```
Comment (adamwojs) [inline SystemInfo/Collector/JsonComposerLockSystemInfoCollector.php:82]: Same comment as above

File: composer.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
         "ocramius/proxy-manager": "~1.0 || ~2.0",
         "symfony/proxy-manager-bridge": "^2.8.40 || ^3.4.11",
         "zetacomponents/system-information": "^1.1.1",
-        "ezsystems/ezplatform-admin-ui": "^1.5@dev"
+        "ezsystems/ezplatform-admin-ui": "dev-ibx-385-responsiveness-if-composer-files-corrupted as 1.5.x-dev"
```
Comment (adamwojs) [inline composer.json:18]: Friendly reminder: revert this change before merge

**Review by adamwojs — COMMENTED** [review]
Also test coverage is missing.

File: composer.json (hunk 4 above)
Comment (konradoboza) [reply composer.json:18]: It's mentioned in the PR description.

**Comment by konradoboza** [comment]
@adamwojs test case added: https://github.com/ezsystems/ez-support-tools/pull/94/commits/13f8bad4d7b94c4f30f6ba61a555afe38cb009ab.

**Review by adamwojs — APPROVED** [review]

File: SystemInfo/Value/InvalidSystemInfo.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/**
+ * File containing the SymfonyKernelSystemInfo class.
```
Comment (ViniTou) [inline SystemInfo/Value/InvalidSystemInfo.php:4]: All those docs are wrong in that matter, I would remove those lines tbh.

**Review by ViniTou — APPROVED** [review]

File: SystemInfo/Value/InvalidSystemInfo.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzSupportToolsBundle\SystemInfo\Value;
+
+use eZ\Publish\API\Repository\Values\ValueObject;
+
+/**
+ * Invalid value for system info used in case of any errors occur while collecting data.
+ */
+class InvalidSystemInfo extends ValueObject implements SystemInfo
```
Comment (Steveb-p) [inline SystemInfo/Value/InvalidSystemInfo.php:16]:
```suggestion
final class InvalidSystemInfo extends ValueObject implements SystemInfo
```

File: SystemInfo/Exception/ComposerFileValidationException.php [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * File containing the ComposerJsonFileNotFoundException class.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzSupportToolsBundle\SystemInfo\Exception;
+
+use Exception;
+
+class ComposerFileValidationException extends Exception
```
Comment (Steveb-p) [inline SystemInfo/Exception/ComposerFileValidationException.php:13]:
```suggestion
final class ComposerFileValidationException extends Exception
```

File: View/SystemInfoViewBuilder.php [hunk 8]
```diff
@@ -37,8 +38,15 @@ public function buildView(array $parameters)
     {
         $collector = $this->getCollector($parameters['systemInfoIdentifier']);
         $view = new SystemInfoView(null, [], $parameters['viewType']);
-        $view->setInfo($collector->collect());
 
+        try {
+            $collectedData = $collector->collect();
+        } catch (\Exception $e) {
```
Comment (Steveb-p) [inline View/SystemInfoViewBuilder.php:44]: I'd introduce an interface for those exceptions, so we do not try to catch **all** exceptions here.

**Review by DominikaK — APPROVED** [review]
+1 for strings.

File: View/SystemInfoViewBuilder.php (hunk 8 above)
Comment (konradoboza) [reply View/SystemInfoViewBuilder.php]: Went for an empty one, feel free to propose something more suitable if you think it's not enough for generalization.

**Review by alongosz — CHANGES_REQUESTED** [review]

File: Tests/SystemInfo/Collector/JsonComposerLockSystemInfoCollectorTest.php [hunk 9]
```diff
@@ -96,4 +97,32 @@ public function testCollectJsonFileNotFound()
         $composerCollectorNotFound = new JsonComposerLockSystemInfoCollector(__DIR__ . '/_fixtures/composer.lock', __DIR__ . '/_fixtures/snafu.json');
         $composerCollectorNotFound->collect();
     }
+
+    /**
+     * @covers \EzSystems\EzSupportToolsBundle\SystemInfo\Collector\JsonComposerLockSystemInfoCollector::collect()
+     */
+    public function testCollectLockFileCorrupted()
```
[the remark this answers is not in the document]
Comment (konradoboza) [reply Tests/SystemInfo/Collector/JsonComposerLockSystemInfoCollectorTest.php]: https://github.com/ezsystems/ez-support-tools/pull/94/commits/f0f8995c1739c8ab7078dbc6c56135d07c75aaed.

**Review by alongosz — APPROVED** [review]

**Review by Steveb-p — APPROVED** [review]

**Comment by konradoboza** [comment]
Removing Admin-UI dependency, Travis lits green - ref: https://travis-ci.com/github/ezsystems/ez-support-tools/builds/226698913.

**Comment by lserwatka** [comment]
You can merge it up.
